compiler/rustc_driver/src/lib.rs

```diff
@@ -1382,3 +1382,56 @@ pub fn main() -> ! {
 
     process::exit(exit_code)
 }
+
+// Configure jemalloc as the `global_allocator` when configured. This is
+// so that we use the sized deallocation apis jemalloc provides
+// (namely `sdallocx`).
+//
+// The symbol overrides documented below are also performed so that we can
+// ensure that we use a consistent allocator across the rustc <-> llvm boundary
+#[cfg(feature = "jemalloc")]
+#[global_allocator]
+static ALLOC: tikv_jemallocator::Jemalloc = tikv_jemallocator::Jemalloc;
+
+// Pull in jemalloc when enabled.
+//
+// Note that we're pulling in a static copy of jemalloc which means that to
+// pull it in we need to actually reference its symbols for it to get
+// linked. The two crates we link to here, std and rustc_driver, are both
+// dynamic libraries. That means to pull in jemalloc we actually need to
+// reference allocation symbols one way or another (as this file is the only
+// object code in the rustc executable).
+#[cfg(feature = "jemalloc")]
+pub fn set_jemalloc() {
```

Comment:
This isn't actually doing anything at runtime, right, only declaring linker symbols? I think you can get rid of the function altogether and make this a module with `cfg(jemalloc)`.

I'm not sure how that interacts with dynamic linking though - I'm worried this will cause rustc_driver to use jemalloc but rustdoc's allocations will still go through the system allocator.